src/global/core/components/animations/animation-abstract.ts

```diff
@@ -0,0 +1,617 @@
+import {
+  AnimationDirection,
+  AnimationFill,
+  AnimationKeyFrames,
+  AnimationLifecycle,
+  InterfaceAnimation,
+  InterfaceAnimationCallbackOptions,
+  InterfaceAnimationInternal,
+  InterfaceAnimationKeyFrame,
+  InterfaceAnimationKeyFrameEdge,
+  InterfaceAnimationOnFinishCallback,
+  InterfaceAnimationPlayOptions
+} from './animation-interface';
+import {
+  addClassToArray,
+  setStyleProperty
+} from './animation-utils';
+
+export abstract class AbstractAnimation implements InterfaceAnimationInternal {
+
+  // eslint-disable-next-line @typescript-eslint/naming-convention
+  protected static readonly _ANIMATION_END_FALLBACK_PADDING_MS = 100;
+
+  public constructor(animationId?: string) {
+    this.id = animationId;
+  }
+
+  public parentAnimation: InterfaceAnimationInternal | undefined;
+  public id: string | undefined;
+  public elements: HTMLElement[] = [];
+  public childAnimations: InterfaceAnimationInternal[] = [];
+
+  protected animationDuration: number | undefined;
+  protected animationKeyframes: AnimationKeyFrames = [];
+  protected cssAnimationsTimerFallback: NodeJS.Timeout;
+  protected finished = false;
+  protected forceDelayValue: number | undefined;
+  protected forceDirectionValue: AnimationDirection | undefined;
+  protected forceDurationValue: number | undefined;
+  protected initialized = false;
+  protected shouldForceLinearEasing = false;
+  protected stylesheets: HTMLElement[] = [];
+  protected webAnimations: any[] = [];
+  protected willComplete = true;
+
+  private _afterAddClasses: string[] = [];
+  private _afterAddReadFunctions: (() => void)[] = [];
+  private _afterAddWriteFunctions: (() => void)[] = [];
+  private _afterRemoveClasses: string[] = [];
+  private _afterStylesValue: { [property: string]: any } = {};
+  private _beforeAddClasses: string[] = [];
+  private _beforeAddReadFunctions: (() => void)[] = [];
+  private _beforeAddWriteFunctions: (() => void)[] = [];
+  private _beforeRemoveClasses: string[] = [];
+  private _beforeStylesValue: { [property: string]: any } = {};
+  private _delay: number | undefined;
+  private _direction: AnimationDirection | undefined;
+  private _easing: string | undefined;
+  private _fill: AnimationFill | undefined;
+  private _iterations: number | undefined;
+  private _numAnimationsRunning = 0;
+  private _onFinishCallbacks: InterfaceAnimationOnFinishCallback[] = [];
+  private _onFinishOneTimeCallbacks: InterfaceAnimationOnFinishCallback[] = [];
+  private _shouldCalculateNumAnimations = true;
+  private _shouldForceSyncPlayback = false;
+
+  public abstract update: (deep: boolean, toggleAnimationName: boolean, step?: number) => InterfaceAnimationInternal;
+
+  protected abstract playImpl: () => void;
```

Comment:
Nitpick; I dislike "Impl" as a suffix. Could we maybe use "Internal" instead? Also applies to the method below.